### ISSUE TITLE
False positive in needless_pass_by_value trait methods

### ISSUE THREAD

**Issue by dtolnay** [rust-lang-nursery/rust-clippy#2208]
```rust
trait FalsePositive {
    fn visit_str(s: &str);
    fn visit_string(s: String) {
        Self::visit_str(&s);
    }
}
```

```
warning: this argument is passed by value, but not consumed in the function body
  --> src/main.rs:37:24
   |
37 |     fn visit_string(s: String) {
   |                        ^^^^^^ help: consider changing the type to: `&str`
   |
   = note: #[warn(needless_pass_by_value)] on by default
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.169/index.html#needless_pass_by_value
```
This is a false positive because the trait gives implementations permission to consume ownership of the string if they want, even though the default implementation does not.

Looks like this is a consequence of #2199. cc @sinkuu
